keystone-init/keystone_init.py

```diff
@@ -316,6 +316,74 @@ def get_or_create_group(client, domain, name):
 
 
 @retry()
+def get_or_create_service(client, name, service_type, description):
+    """Get service or create it if doesn't exist.
+
+    :type client: keystoneclient.v3.client.Client
+    :type name: str
+    :type service_type: str
+    :type description: str
+    :return:
+    :rtype: keystoneclient.v3.services.Service
+    """
+    services = client.services.list()
+    service_names = {service.name: service for service in services}
+
+    if name in service_names.keys():
+        return service_names[name]
+    else:
+        logger.info('creating new service %s of %s type', name, service_type)
+        service = client.services.create(
+            name=name,
+            type=service_type,
+            description=description,
+        )
+        logger.debug('created service %r', service)
+
+    return service
+
+
+@retry()
+def get_or_create_endpoint(client, service, url, interface, region):
+    """Get endpoint or create it if doesn't exist.
+
+    :type client: keystoneclient.v3.client.Client
+    :type service: keystoneclient.v3.services.Service
+    :type url: str
+    :type interface: dict[str, str]
+    :type region: str
+    :return:
+    :rtype: keystoneclient.v3.endpoints.Endpoint
+    """
+    endpoints = client.endpoints.list()
+
+    for endpoint in endpoints:
+        logger.debug('existing endpoint %r', endpoint)
+        if endpoint.service_id == service.id:
+            if endpoint.url == url:
+                if endpoint.interface == interface['name']:
+                    return endpoint
+            else:
+                client.endpoints.delete(endpoint)
```

Comment:
I'm not sure deleting endpoints that already exist is desirable, but if that's really the behavior we want we should at least log a warning about it.

Generally I'd prefer to create or update resources when the preload input specifies but leave everything else alone, unless there's some cases where an extra endpoint could cause issues. So far we've been able to get away without deleting anything, which is a lot less scary IMO.

Reply:
Changed to only updating existing endpoints